crates/ra_lsp_server/src/cargo_check.rs

```diff
@@ -0,0 +1,1316 @@
+                    };
+
+                let MappedRustDiagnostic { location, diagnostic, suggested_fixes } = map_result;
+                let file_uri = location.uri.clone();
+
+                if !suggested_fixes.is_empty() {
+                    for suggested_fix in suggested_fixes {
+                        self.shared
+                            .write()
+                            .add_suggested_fix_for_diagnostic(suggested_fix, &diagnostic);
+                    }
+                }
+                self.shared.write().add_diagnostic(file_uri, diagnostic);
+
+                task_send.send(CheckTask::Update(location.uri)).unwrap();
+            }
+
+            CheckEvent::Msg(Message::BuildScriptExecuted(_msg)) => {}
+            CheckEvent::Msg(Message::Unknown) => {}
+        }
+    }
+}
+
+/// WatchThread exists to wrap around the communication needed to be able to
+/// run `cargo check` without blocking. Currently the Rust standard library
+/// doesn't provide a way to read sub-process output without blocking, so we
+/// have to wrap sub-processes output handling in a thread and pass messages
+/// back over a channel.
+struct WatchThread {
+    message_recv: Receiver<CheckEvent>,
+    cancel_send: Sender<()>,
+}
+
+enum CheckEvent {
+    Begin,
+    Msg(cargo_metadata::Message),
+    End,
+}
+
+impl WatchThread {
+    fn new(options: &Options, workspace_root: &PathBuf) -> WatchThread {
+        let mut args: Vec<String> = vec![
+            options.cargo_watch_command.clone(),
+            "--message-format=json".to_string(),
+            "--manifest-path".to_string(),
+            format!("{}/Cargo.toml", workspace_root.to_string_lossy()),
+        ];
+        if options.cargo_watch_all_targets {
+            args.push("--all-targets".to_string());
+        }
+        args.extend(options.cargo_watch_args.iter().cloned());
+
+        let (message_send, message_recv) = unbounded();
+        let (cancel_send, cancel_recv) = unbounded();
+        let enabled = options.cargo_watch_enable;
+        std::thread::spawn(move || {
+            if !enabled {
+                return;
+            }
+
+            let mut command = Command::new("cargo")
+                .args(&args)
+                .stdout(Stdio::piped())
+                .stderr(Stdio::null())
+                .spawn()
+                .expect("couldn't launch cargo");
+
+            message_send.send(CheckEvent::Begin).unwrap();
+            for message in cargo_metadata::parse_messages(command.stdout.take().unwrap()) {
+                match cancel_recv.try_recv() {
+                    Ok(()) | Err(TryRecvError::Disconnected) => {
+                        command.kill().expect("couldn't kill command");
+                    }
+                    Err(TryRecvError::Empty) => (),
+                }
+
+                message_send.send(CheckEvent::Msg(message.unwrap())).unwrap();
+            }
+            message_send.send(CheckEvent::End).unwrap();
+        });
```

Comment:
I think there are two problems with cancellation here:

* first, we drop `JoinHandle` on the floor. This usually is a bad idea, because then, for example, you might get interference in tests.
* second, `command` might take arbitrary long time to run, but we need to be able to interrupt it when:
  * we recieve a new change (we need to kill and restart `cargo check` then)
  * the server itself is shutdown (we need to shutdown quickly, without waiting for check to finish, and without leaving orphan processes behind)

To handle the first problem, you can use https://github.com/matklad/jod-thread (it might be annoying to setup, as "perfectly fine" code might start to block, but it should lead to a more robust system in the end).

To handle the second problem, I expect we need some kind of a wrapper around `std::process::Child` which calls `Child::kill` in `Drop`.